File: Python/Stack/1209. Remove All Adjacent Duplicates in String II.py
class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        stack = []
        count = 1
        for c in s:
            count = 1 if (not stack or stack[-1][0]!=c) else stack[-1][1]+1
            stack.append((c, count))
            if stack[-1][1] >= k:
                for _ in range(k):
                    stack.pop()
        res = ""
        for c, count in stack:
            res += c
        return res

File: C++/Stack/1209. Remove All Adjacent Duplicates in String II.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) {
        deque<pair<char, int>> dq; // {letter, count}
        for (char c : s) {
            if (!dq.empty() && c == dq.back().first) {
                if (dq.back().second >= k-1) {
                    dq.pop_back();
                }
                else {
                    dq.back().second++;
                }
            }
            else {
                dq.push_back({c, 1});
            }
        }
        string ans = "";
        while (!dq.empty()) {
            for (int i = 0; i < dq.front().second; i++) {
                ans += dq.front().first;
            }
            dq.pop_front();
        }
        return ans;
    }
};